### PR TITLE
[Datasets] Fix Dataset.count() to not trigger execution

### PR DESCRIPTION
**Pull request by c21**
Signed-off-by: Cheng Su <scnju13@gmail.com>

## Why are these changes needed?

This is to fix the issue found in nightly test, that `Dataset.count()` trigger unexpected execution, instead of calculating the number of rows from metadata directly. This PR is to fix the issue, reran the nightly test and it works as expected now.

## Checks

- [x] I've signed off every commit(by using the -s flag, i.e., `git commit -s`) in this PR.
- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [ ] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

**Review by ericl — APPROVED** [review]

File: python/ray/data/tests/test_dataset.py [hunk 1]
```diff
@@ -1487,6 +1487,15 @@ def test_schema_lazy(ray_start_regular_shared):
     assert ds._plan.execute()._num_computed() == 1
 
 
+def test_count_lazy(ray_start_regular_shared):
+    ds = ray.data.range(100, parallelism=10)
+    # We do not kick off the read task by default.
+    assert ds._plan._in_blocks._num_computed() == 0
+    assert ds.count() == 100
+    # Getting number of rows should not trigger execution of any read tasks.
```
Comment (jianoaix) [inline python/ray/data/tests/test_dataset.py:1495]: Is this always true for other ways of dataset creation, like read_csv?

File: python/ray/data/_internal/plan.py [hunk 2]
```diff
@@ -399,6 +399,11 @@ def meta_count(self) -> Optional[int]:
         """
         if self._stages_after_snapshot:
             return None
+        elif self._in_blocks is not None and self._snapshot_blocks is None:
+            # If the plan only has input blocks, we execute it, so snapshot has output.
+            # This applies to newly created dataset. For example, initial dataset from
+            # read, and output datasets of Dataset.split().
+            self.execute()
         # Snapshot is now guaranteed to be the output of the final stage or None.
```
Comment (jianoaix) [inline python/ray/data/_internal/plan.py:407]: This comment seems not quite right, if the input blocks is lazy blocklist, after a `self.execute()` (without force read), this snapshot is still a lazy list, which is not a final output.

File: python/ray/data/tests/test_dataset.py (hunk 1 above)
Comment (c21) [reply python/ray/data/tests/test_dataset.py:1495]: yeah I think so, they all use same code path - `read_datasource()`.

File: python/ray/data/_internal/plan.py (hunk 2 above)
Comment (c21) [reply python/ray/data/_internal/plan.py:407]: I can remove the comment if it's confusing, but it just meant to say `_snapshot_blocks` is the latest block to get output from I guess.

File: python/ray/data/tests/test_dataset.py (hunk 1 above)
Comment (c21) [reply python/ray/data/tests/test_dataset.py]: Updated comment.

File: python/ray/data/_internal/plan.py (hunk 2 above)
Comment (c21) [reply python/ray/data/_internal/plan.py]: Updated comment.

**Review by jianoaix — APPROVED** [review]

**Review by clarkzinzow — APPROVED** [review]

**Pull request merged**